depends: Upgrade Qt to 5.9.6

sedited opened the pull request:
With the introduction of Ubuntu 18.04 (Bionic Beaver) modern versions of gcc and mingw that allow cross compilation of versions of Qt greater than 5.8 are now readily available. This pull requests upgrades the Qt depends recipe from Qt 5.7.1 to Qt 5.9.6. Qt 5.9.x is the current LTS version and should be supported by Qt until 2020.

MarcoFalke commented:
According to https://doc.qt.io/qt-5.10/supported-platforms-and-configurations.html#qt-5-9 gcc4.8 is supported. Also, I could build your receipt on ubuntu trusty, which comes with gcc4.8. I think you can get rid of the optional flag to compile qt5.9?

TheCharlatan commented:
@MarcoFalke The compiler version for linux native compilation is not a problem. It is though as soon as you are trying to cross compile to windows (which requires Mingw5.3). 
I originally did not want to open the pr for this reason, since it would mean that the current travis tests would be broken because of the windows build requirement. I heard though that @theuni is working on getting the entire toolchain on depends, so I decided to publish the patch nevertheless to prevent duplicate work and enable other developers to already compile bitcoin static with a modern Qt version. 

The macos build needs some further investigation. It fails upon creating qmake and the current qmake patch does not seem to work. Upon some investigation, I saw that multiple people had to do some manual property setting on xcode to make the build work. I am not sure if this is related though.

MarcoFalke commented:
> cross compile to windows (which requires Mingw5.3)

Hmm. Imo we could drop support for cross-compile to windows on trusty and just require xenial or above.

MarcoFalke commented:
Could do a rebase since #12946 is merged?

TheCharlatan commented:
Rebased.

fanquake commented:
@TheCharlatan Thanks for the work so far, however did you see my comment above? I think this would be better as a straight 5.7 -> 5.9 upgrade, instead of adding an additional version of Qt to depends. We would like to use Qt 5.9+ to build the next release anyways.

Regardless of the that this needs a few fixes.

There are currently [whitespace linting errors](https://travis-ci.org/bitcoin/bitcoin/jobs/366832841#L794).

A linux Travis build is [currently failing](https://travis-ci.org/bitcoin/bitcoin/jobs/366832847):
[linux-raw-log.txt](https://github.com/bitcoin/bitcoin/files/1913637/linux-raw-log.txt)

```
configure: error: Package requirements (Qt5PlatformSupport) were not met:
No package 'Qt5PlatformSupport' found
Consider adjusting the PKG_CONFIG_PATH environment variable if you
installed software in a non-standard prefix.
Alternatively, you may set the environment variables QTPLATFORM_CFLAGS
and QTPLATFORM_LIBS to avoid the need to call pkg-config.
See the pkg-config man page for more details.
This file contains any messages produced by compilers while
running configure, to aid debugging if configure makes a mistake.
It was created by Bitcoin Core configure 0.16.99, which was
generated by GNU Autoconf 2.69.  Invocation command line was
  $ ../configure --cache-file=config.cache --disable-dependency-tracking --prefix=/home/travis/build/bitcoin/bitcoin/depends/x86_64-unknown-linux-gnu --bindir=/home/travis/build/bitcoin/bitcoin/out/12971/27461.7-x86_64-unknown-linux-gnu/bin --libdir=/home/travis/build/bitcoin/bitcoin/out/12971/27461.7-x86_64-unknown-linux-gnu/lib --enable-glibc-back-compat --enable-reduce-exports
```

The macOS build would also need to be [fixed/updated](https://travis-ci.org/bitcoin/bitcoin/jobs/366832848):
[macOS-raw-log.txt](https://github.com/bitcoin/bitcoin/files/1913643/macOS-raw-log.txt)

```
Configuring qt59...
Creating qmake...
make[1]: Entering directory `/home/travis/build/bitcoin/bitcoin/depends/work/build/x86_64-apple-darwin11/qt59/5.9.4-01715d30425/qtbase/qmake'
make[1]: Leaving directory `/home/travis/build/bitcoin/bitcoin/depends/work/build/x86_64-apple-darwin11/qt59/5.9.4-01715d30425/qtbase/qmake'
Info: creating cache file /home/travis/build/bitcoin/bitcoin/depends/work/build/x86_64-apple-darwin11/qt59/5.9.4-01715d30425/qtbase/.qmake.cache
Command line: -bindir /home/travis/build/bitcoin/bitcoin/depends/x86_64-apple-darwin11/native/bin -c++std c++11 -confirm-license -dbus-runtime -hostprefix /home/travis/build/bitcoin/bitcoin/depends/x86_64-apple-darwin11/native -no-cups -no-egl -no-eglfs -no-freetype -no-gif -no-glib -no-icu -no-iconv -no-kms -no-linuxfb -no-libudev -no-mtdev -no-openvg -no-reduce-relocations -no-qml-debug -no-sql-db2 -no-sql-ibase -no-sql-oci -no-sql-tds -no-sql-mysql -no-sql-odbc -no-sql-psql -no-sql-sqlite -no-sql-sqlite2 -no-use-gold-linker -nomake examples -nomake tests -opensource -openssl-linked -optimized-qmake -pch -pkg-config -prefix /home/travis/build/bitcoin/bitcoin/depends/x86_64-apple-darwin11 -qt-libpng -qt-libjpeg -qt-pcre -system-zlib -static -silent -v -release -xplatform macx-clang-linux -device-option MAC_SDK_PATH=/home/travis/build/bitcoin/bitcoin/depends/SDKs/MacOSX10.11.sdk -device-option MAC_SDK_VERSION=10.11 -device-option CROSS_COMPILE=x86_64-apple-darwin11- -device-option MAC_MIN_VERSION=10.8 -device-option MAC_TARGET=x86_64-apple-darwin11 -device-option MAC_LD64_VERSION=253.9
Project ERROR: Could not resolve SDK --show-sdk-path for 'macosx'
make: *** [/home/travis/build/bitcoin/bitcoin/depends/work/build/x86_64-apple-darwin11/qt59/5.9.4-01715d30425/qtbase/.stamp_configured] Error 3
make: Leaving directory `/home/travis/build/bitcoin/bitcoin/depends'
```

fanquake commented on depends/packages/packages.mk:
```diff
@@ -3,11 +3,19 @@ packages:=boost openssl libevent zeromq
+qt_x86_64_linux_packages:=qt59 expat dbus libxcb xcb_proto libXau xproto freetype fontconfig libX11 xextproto libXext xtrans
 qt_i686_linux_packages:=$(qt_x86_64_linux_packages)
 
-qt_darwin_packages=qt
```
This looks like it's building QT 5.9 in either case?

TheCharlatan commented:
@fanquake thanks for the review. 
If I were to change this to a straight upgrade, how would travis, or gitian windows builds be done?

MarcoFalke commented:
Our gitian descriptors will be switched to bionic soon: #12511. Similarly, we can switch travis to xenial. Those will happen in separate pull request, so don't worry about them.

laanwj commented:
Travis fails on linux 64 bit with:
```
checking for QTPLATFORM... no
configure: error: Package requirements (Qt5PlatformSupport) were not met:
No package 'Qt5PlatformSupport' found
```

And OSX:
```
Project ERROR: Could not resolve SDK --show-sdk-path for 'macosx'
```

MarcoFalke commented:
See also #13215

fanquake commented:
@TheCharlatan Can you fixup any trailing whitespace so that Travis will run: https://travis-ci.org/bitcoin/bitcoin/jobs/379496754

MarcoFalke commented on depends/README.md:
```diff
@@ -12,6 +12,10 @@ For example:
 
+To build the more recent Qt5.9.4, you can build depends with
+
+  make HOST=host-platform-triple QT_59=1
```
Can remove this hunk and all other references to qt 5.9?

MarcoFalke commented on depends/packages/qt.mk:
```diff
@@ -75,12 +64,10 @@ $(package)_config_opts += -qt-libpng
 $(package)_config_opts += -qt-pcre
 $(package)_config_opts += -system-zlib
-$(package)_config_opts += -reduce-exports
+#$(package)_config_opts += -reduce-exports
```
nit: Either remove or replace with proper comment/reasoning.

MarcoFalke commented on doc/build-windows.md:
```diff
@@ -117,6 +117,20 @@ Once the source code is ready the build steps are below.
+    ./autogen.sh # not required when building from tarball
+    CONFIG_SITE=$PWD/depends/x86_64-w64-mingw32/share/config.site ./configure --prefix=/ --enable-qt59
+    make
+
```
This whole section is added in #13246, so feel free to remove this to avoid merge conflicts.

ken2812221 commented:
This should fix the problem. Patch from [QTBUG-67286](https://bugreports.qt.io/browse/QTBUG-67286)
```diff
diff --git a/depends/packages/qt.mk b/depends/packages/qt.mk
index 090635651..725daf2f5 100644
--- a/depends/packages/qt.mk
+++ b/depends/packages/qt.mk
@@ -8,7 +8,7 @@ $(package)_dependencies=openssl zlib
 $(package)_linux_dependencies=freetype fontconfig libxcb libX11 xproto libXext
 $(package)_build_subdir=qtbase
 $(package)_qt_libs=corelib network widgets gui plugins testlib concurrent xml
-$(package)_patches=fix_qt_pkgconfig.patch mac-qmake.conf
+$(package)_patches=fix_qt_pkgconfig.patch mac-qmake.conf fix_configure_mac.patch
 
 $(package)_qttranslations_file_name=qttranslations-$($(package)_suffix)
 $(package)_qttranslations_sha256_hash=e5cf81f88cc6811166ea631e6b7faa869a578a910dce61e951f98ff83f27bad6
@@ -130,7 +130,8 @@ define $(package)_preprocess_cmds
   sed -i.old "s|QMAKE_CFLAGS            = |!host_build: QMAKE_CFLAGS            = $($(package)_cflags) $($(package)_cppflags) |" qtbase/mkspecs/win32-g++/qmake.conf && \
   sed -i.old "s|QMAKE_LFLAGS            = |!host_build: QMAKE_LFLAGS            = $($(package)_ldflags) |" qtbase/mkspecs/win32-g++/qmake.conf && \
   sed -i.old "s|QMAKE_CXXFLAGS          = |!host_build: QMAKE_CXXFLAGS            = $($(package)_cxxflags) $($(package)_cppflags) |" qtbase/mkspecs/win32-g++/qmake.conf &&\
-  patch -p1 -i $($(package)_patch_dir)/fix_qt_pkgconfig.patch
+  patch -p1 -i $($(package)_patch_dir)/fix_qt_pkgconfig.patch && \
+  patch -p1 -i $($(package)_patch_dir)/fix_configure_mac.patch
 endef
 
 define $(package)_config_cmds
diff --git a/depends/patches/qt/fix_configure_mac.patch b/depends/patches/qt/fix_configure_mac.patch
new file mode 100644
index 000000000..0d7dd647d
--- /dev/null
+++ b/depends/patches/qt/fix_configure_mac.patch
@@ -0,0 +1,50 @@
+--- old/qtbase/mkspecs/features/mac/sdk.prf	2018-02-08 10:24:48.000000000 -0800
++++ new/qtbase/mkspecs/features/mac/sdk.prf	2018-03-23 10:38:56.000000000 -0700
+@@ -8,21 +8,21 @@
+ defineReplace(xcodeSDKInfo) {
+     info = $$1
+     equals(info, "Path"): \
+-        info = --show-sdk-path
++        infoarg = --show-sdk-path
+     equals(info, "PlatformPath"): \
+-        info = --show-sdk-platform-path
++        infoarg = --show-sdk-platform-path
+     equals(info, "SDKVersion"): \
+-        info = --show-sdk-version
++        infoarg = --show-sdk-version
+     sdk = $$2
+     isEmpty(sdk): \
+         sdk = $$QMAKE_MAC_SDK
+ 
+     isEmpty(QMAKE_MAC_SDK.$${sdk}.$${info}) {
+-        QMAKE_MAC_SDK.$${sdk}.$${info} = $$system("/usr/bin/xcrun --sdk $$sdk $$info 2>/dev/null")
++        QMAKE_MAC_SDK.$${sdk}.$${info} = $$system("/usr/bin/xcrun --sdk $$sdk $$infoarg 2>/dev/null")
+         # --show-sdk-platform-path won't work for Command Line Tools; this is fine
+         # only used by the XCTest backend to testlib
+-        isEmpty(QMAKE_MAC_SDK.$${sdk}.$${info}):if(!isEmpty(QMAKE_XCODEBUILD_PATH)|!equals(info, "--show-sdk-platform-path")): \
+-            error("Could not resolve SDK $$info for \'$$sdk\'")
++        isEmpty(QMAKE_MAC_SDK.$${sdk}.$${info}):if(!isEmpty(QMAKE_XCODEBUILD_PATH)|!equals(infoarg, "--show-sdk-platform-path")): \
++            error("Could not resolve SDK $$info for \'$$sdk\' using $$infoarg")
+         cache(QMAKE_MAC_SDK.$${sdk}.$${info}, set stash, QMAKE_MAC_SDK.$${sdk}.$${info})
+     }
+
+--- old/qtbase/configure	2018-02-08 10:24:48.000000000 -0800
++++ new/qtbase/configure	2018-03-23 05:42:29.000000000 -0700
+@@ -232,8 +232,13 @@
+ 
+     sdk=$(getSingleQMakeVariable "QMAKE_MAC_SDK" "$1")
+     if [ -z "$sdk" ]; then echo "QMAKE_MAC_SDK must be set when building on Mac" >&2; exit 1; fi
+-    sysroot=$(/usr/bin/xcrun --sdk $sdk --show-sdk-path 2>/dev/null)
+-    if [ -z "$sysroot" ]; then echo "Failed to resolve SDK path for '$sdk'" >&2; exit 1; fi
++    sysroot=$(getSingleQMakeVariable "QMAKE_MAC_SDK_PATH" "$1")
++
++    echo "sysroot pre-configured as $sysroot";
++    if [ -z "$sysroot" ]; then
++       sysroot=$(/usr/bin/xcrun --sdk $sdk --show-sdk-path 2>/dev/null)
++       if [ -z "$sysroot" ]; then echo "Failed to resolve SDK path for '$sdk'" >&2; exit 1; fi
++    fi
+ 
+     case "$sdk" in
+         macosx*)
+
+
diff --git a/depends/patches/qt/mac-qmake.conf b/depends/patches/qt/mac-qmake.conf
index ca70d30b1..337d0eb9c 100644
--- a/depends/patches/qt/mac-qmake.conf
+++ b/depends/patches/qt/mac-qmake.conf
@@ -14,6 +14,7 @@ QMAKE_MAC_SDK.macosx.Path = $${MAC_SDK_PATH}
 QMAKE_MAC_SDK.macosx.platform_name = macosx
 QMAKE_MAC_SDK.macosx.SDKVersion = $${MAC_SDK_VERSION}
 QMAKE_MAC_SDK.macosx.PlatformPath = /phony
+QMAKE_APPLE_DEVICE_ARCHS=x86_64
 !host_build: QMAKE_CFLAGS += -target $${MAC_TARGET}
 !host_build: QMAKE_OBJECTIVE_CFLAGS += $$QMAKE_CFLAGS
 !host_build: QMAKE_CXXFLAGS += $$QMAKE_CFLAGS

ken2812221 commented:
This additional patch should fix all Mac and Windows build:
```diff
diff --git a/build-aux/m4/bitcoin_qt.m4 b/build-aux/m4/bitcoin_qt.m4
index 28446a3cf..bb5d26c2d 100644
--- a/build-aux/m4/bitcoin_qt.m4
+++ b/build-aux/m4/bitcoin_qt.m4
@@ -292,7 +292,7 @@ AC_DEFUN([_BITCOIN_QT_CHECK_QT5],[
       #endif
     ]],
     [[
-      #if QT_VERSION < 0x050000 OR QT_VERSION_MAJOR < 5
+      #if QT_VERSION < 0x050000 || QT_VERSION_MAJOR < 5
       choke
       #endif
     ]])],
@@ -303,7 +303,7 @@ AC_DEFUN([_BITCOIN_QT_CHECK_QT5],[
 dnl Internal. Check if the included version of Qt is greater than Qt57.
 dnl Requires: INCLUDES must be populated as necessary.
 dnl Output: bitcoin_cv_qt5=yes|no
-AC_DEFUN([_BITCOIN_QT_CHECK_QT5],[
+AC_DEFUN([_BITCOIN_QT_CHECK_QT58],[
   AC_CACHE_CHECK(for > Qt 5.7, bitcoin_cv_qt58,[
   AC_COMPILE_IFELSE([AC_LANG_PROGRAM([[
       #include <QtCore/qconfig.h>
@@ -384,12 +384,16 @@ AC_DEFUN([_BITCOIN_QT_FIND_STATIC_PLUGINS],[
        if test x$bitcoin_cv_qt58 = xno; then
          PKG_CHECK_MODULES([QTPLATFORM], [Qt5PlatformSupport], [QT_LIBS="$QTPLATFORM_LIBS $QT_LIBS"])
        else
-         PKG_CHECK_MODULES([QTPLATFORM], [Qt5FontDatabaseSupport], [QT_LIBS="$QTFONTDATABASE_LIBS $QT_LIBS"])
-         PKG_CHECK_MODULES([QTPLATFORM], [Qt5EventDispatcherSupport], [QT_LIBS="$QTEVENTDISPATCHER_LIBS $QT_LIBS"])
-         PKG_CHECK_MODULES([QTPLATFORM], [Qt5ThemeSupport], [QT_LIBS="$QTTHEME_LIBS $QT_LIBS"])
-         PKG_CHECK_MODULES([QTPLATFORM], [Qt5DeviceDiscoverySupport], [QT_LIBS="$QTDEVICEDISCOVERY_LIBS $QT_LIBS"])
-         PKG_CHECK_MODULES([QTPLATFORM], [Qt5AccessibilitySupport], [QT_LIBS="$QTACCESSIBILITY_LIBS $QT_LIBS"])
-         PKG_CHECK_MODULES([QTPLATFORM], [Qt5FbSupport], [QT_LIBS="$QTFB_LIBS $QT_LIBS"])
+         PKG_CHECK_MODULES([QTPLATFORM], [Qt5FontDatabaseSupport], [QT_LIBS="-lQt5FontDatabaseSupport $QT_LIBS"])
+         PKG_CHECK_MODULES([QTPLATFORM], [Qt5EventDispatcherSupport], [QT_LIBS="-lQt5EventDispatcherSupport $QT_LIBS"])
+         PKG_CHECK_MODULES([QTPLATFORM], [Qt5ThemeSupport], [QT_LIBS="-lQt5ThemeSupport $QT_LIBS"])
+         PKG_CHECK_MODULES([QTPLATFORM], [Qt5DeviceDiscoverySupport], [QT_LIBS="-lQt5DeviceDiscoverySupport $QT_LIBS"])
+         PKG_CHECK_MODULES([QTPLATFORM], [Qt5AccessibilitySupport], [QT_LIBS="-lQt5AccessibilitySupport $QT_LIBS"])
+         PKG_CHECK_MODULES([QTPLATFORM], [Qt5FbSupport], [QT_LIBS="-lQt5FbSupport $QT_LIBS"])
+         PKG_CHECK_MODULES([QTPLATFORM], [Qt5ClipboardSupport], [QT_LIBS="-lQt5ClipboardSupport $QT_LIBS"])
+         PKG_CHECK_MODULES([QTPLATFORM], [Qt5GraphicsSupport], [QT_LIBS="-lQt5GraphicsSupport $QT_LIBS"])
+         PKG_CHECK_MODULES([QTPLATFORM], [Qt5OpenGL], [QT_LIBS="-lQt5OpenGL $QT_LIBS"])
+         PKG_CHECK_MODULES([QTPLATFORM], [Qt5CglSupport], [QT_LIBS="-lQt5CglSupport $QT_LIBS"])
        fi
        if test "x$TARGET_OS" = xlinux; then
          PKG_CHECK_MODULES([X11XCB], [x11-xcb], [QT_LIBS="$X11XCB_LIBS $QT_LIBS"])
@@ -410,7 +414,7 @@ AC_DEFUN([_BITCOIN_QT_FIND_STATIC_PLUGINS],[
                #endif
              ]],
              [[
-               #if QT_VERSION < 0x050600 OR QT_VERSION_MINOR < 6
+               #if QT_VERSION < 0x050600 || QT_VERSION_MINOR < 6
                choke
                #endif
              ]])],
@@ -519,6 +523,7 @@ AC_DEFUN([_BITCOIN_QT_FIND_LIBS_WITHOUT_PKGCONFIG],[
   BITCOIN_QT_CHECK([
     if test "x$bitcoin_qt_want_version" = xauto; then
       _BITCOIN_QT_CHECK_QT5
+      _BITCOIN_QT_CHECK_QT58
     fi
     if test "x$bitcoin_cv_qt5" = xyes || test "x$bitcoin_qt_want_version" = xqt5; then
       QT_LIB_PREFIX=Qt5
diff --git a/depends/packages/qt.mk b/depends/packages/qt.mk
index e4a1f3751..884a02c2b 100644
--- a/depends/packages/qt.mk
+++ b/depends/packages/qt.mk
@@ -8,7 +8,7 @@ $(package)_dependencies=openssl zlib
 $(package)_linux_dependencies=freetype fontconfig libxcb libX11 xproto libXext
 $(package)_build_subdir=qtbase
 $(package)_qt_libs=corelib network widgets gui plugins testlib concurrent xml
-$(package)_patches=fix_qt_pkgconfig.patch mac-qmake.conf fix_configure_mac.patch
+$(package)_patches=fix_qt_pkgconfig.patch mac-qmake.conf fix_configure_mac.patch fix_no_printer.patch
 
 $(package)_qttranslations_file_name=qttranslations-$($(package)_suffix)
 $(package)_qttranslations_sha256_hash=e5cf81f88cc6811166ea631e6b7faa869a578a910dce61e951f98ff83f27bad6
@@ -67,6 +67,8 @@ $(package)_config_opts += -system-zlib
 $(package)_config_opts += -static
 $(package)_config_opts += -silent
 $(package)_config_opts += -v
+$(package)_config_opts += -no-feature-printer
+$(package)_config_opts += -no-feature-printdialog
 
 ifneq ($(build_os),darwin)
 $(package)_config_opts_darwin = -xplatform macx-clang-linux
@@ -130,7 +132,8 @@ define $(package)_preprocess_cmds
   sed -i.old "s|QMAKE_LFLAGS            = |!host_build: QMAKE_LFLAGS            = $($(package)_ldflags) |" qtbase/mkspecs/win32-g++/qmake.conf && \
   sed -i.old "s|QMAKE_CXXFLAGS          = |!host_build: QMAKE_CXXFLAGS            = $($(package)_cxxflags) $($(package)_cppflags) |" qtbase/mkspecs/win32-g++/qmake.conf &&\
   patch -p1 -i $($(package)_patch_dir)/fix_qt_pkgconfig.patch &&\
-  patch -p1 -i $($(package)_patch_dir)/fix_configure_mac.patch
+  patch -p1 -i $($(package)_patch_dir)/fix_configure_mac.patch &&\
+  patch -p1 -i $($(package)_patch_dir)/fix_no_printer.patch
 endef
 
 define $(package)_config_cmds
diff --git a/depends/patches/qt/fix_no_printer.patch b/depends/patches/qt/fix_no_printer.patch
new file mode 100644
index 000000000..1cbd70ed4
--- /dev/null
+++ b/depends/patches/qt/fix_no_printer.patch
@@ -0,0 +1,19 @@
+--- x/qtbase/src/plugins/platforms/cocoa/qprintengine_mac_p.h
++++ y/qtbase/src/plugins/platforms/cocoa/qprintengine_mac_p.h
+@@ -52,6 +52,7 @@
+ //
+ 
+ #include <QtCore/qglobal.h>
++#include <qpa/qplatformprintdevice.h>
+ 
+ #ifndef QT_NO_PRINTER
+ 
+--- x/qtbase/src/plugins/plugins.pro
++++ y/qtbase/src/plugins/plugins.pro
+@@ -8,6 +8,3 @@ qtHaveModule(gui) {
+     qtConfig(imageformatplugin): SUBDIRS *= imageformats
+     !android:qtConfig(library): SUBDIRS *= generic
+ }
+-
+-!winrt:qtHaveModule(printsupport): \
+-    SUBDIRS += printsupport

ken2812221 commented:
You committed the wrong patch, it's not `fix-cocoahelpers-macos.patch` but `fix_no_printer.patch`

ken2812221 commented:
Tested ACK 51a5430 on Windows

ken2812221 commented:
Sorry, I did not test this on Linux. Please apply this patch to allow static link Qt on Linux.
```diff
diff --git a/build-aux/m4/bitcoin_qt.m4 b/build-aux/m4/bitcoin_qt.m4
index bb5d26c2d..80f8e4bfb 100644
--- a/build-aux/m4/bitcoin_qt.m4
+++ b/build-aux/m4/bitcoin_qt.m4
@@ -384,16 +384,12 @@ AC_DEFUN([_BITCOIN_QT_FIND_STATIC_PLUGINS],[
        if test x$bitcoin_cv_qt58 = xno; then
          PKG_CHECK_MODULES([QTPLATFORM], [Qt5PlatformSupport], [QT_LIBS="$QTPLATFORM_LIBS $QT_LIBS"])
        else
-         PKG_CHECK_MODULES([QTPLATFORM], [Qt5FontDatabaseSupport], [QT_LIBS="-lQt5FontDatabaseSupport $QT_LIBS"])
-         PKG_CHECK_MODULES([QTPLATFORM], [Qt5EventDispatcherSupport], [QT_LIBS="-lQt5EventDispatcherSupport $QT_LIBS"])
-         PKG_CHECK_MODULES([QTPLATFORM], [Qt5ThemeSupport], [QT_LIBS="-lQt5ThemeSupport $QT_LIBS"])
-         PKG_CHECK_MODULES([QTPLATFORM], [Qt5DeviceDiscoverySupport], [QT_LIBS="-lQt5DeviceDiscoverySupport $QT_LIBS"])
-         PKG_CHECK_MODULES([QTPLATFORM], [Qt5AccessibilitySupport], [QT_LIBS="-lQt5AccessibilitySupport $QT_LIBS"])
-         PKG_CHECK_MODULES([QTPLATFORM], [Qt5FbSupport], [QT_LIBS="-lQt5FbSupport $QT_LIBS"])
-         PKG_CHECK_MODULES([QTPLATFORM], [Qt5ClipboardSupport], [QT_LIBS="-lQt5ClipboardSupport $QT_LIBS"])
-         PKG_CHECK_MODULES([QTPLATFORM], [Qt5GraphicsSupport], [QT_LIBS="-lQt5GraphicsSupport $QT_LIBS"])
-         PKG_CHECK_MODULES([QTPLATFORM], [Qt5OpenGL], [QT_LIBS="-lQt5OpenGL $QT_LIBS"])
-         PKG_CHECK_MODULES([QTPLATFORM], [Qt5CglSupport], [QT_LIBS="-lQt5CglSupport $QT_LIBS"])
+         PKG_CHECK_MODULES([QTFONTDATABASE], [Qt5FontDatabaseSupport], [QT_LIBS="-lQt5FontDatabaseSupport $QT_LIBS"])
+         PKG_CHECK_MODULES([QTEVENTDISPATCHER], [Qt5EventDispatcherSupport], [QT_LIBS="-lQt5EventDispatcherSupport $QT_LIBS"])
+         PKG_CHECK_MODULES([QTTHEME], [Qt5ThemeSupport], [QT_LIBS="-lQt5ThemeSupport $QT_LIBS"])
+         PKG_CHECK_MODULES([QTDEVICEDISCOVERY], [Qt5DeviceDiscoverySupport], [QT_LIBS="-lQt5DeviceDiscoverySupport $QT_LIBS"])
+         PKG_CHECK_MODULES([QTACCESSIBILITY], [Qt5AccessibilitySupport], [QT_LIBS="-lQt5AccessibilitySupport $QT_LIBS"])
+         PKG_CHECK_MODULES([QTFB], [Qt5FbSupport], [QT_LIBS="-lQt5FbSupport $QT_LIBS"])
        fi
        if test "x$TARGET_OS" = xlinux; then
          PKG_CHECK_MODULES([X11XCB], [x11-xcb], [QT_LIBS="$X11XCB_LIBS $QT_LIBS"])
@@ -401,7 +397,9 @@ AC_DEFUN([_BITCOIN_QT_FIND_STATIC_PLUGINS],[
            PKG_CHECK_MODULES([QTXCBQPA], [Qt5XcbQpa], [QT_LIBS="$QTXCBQPA_LIBS $QT_LIBS"])
          fi
        elif test "x$TARGET_OS" = xdarwin; then
-         PKG_CHECK_MODULES([QTPRINT], [Qt5PrintSupport], [QT_LIBS="$QTPRINT_LIBS $QT_LIBS"])
+         PKG_CHECK_MODULES([QTCLIPBOARD], [Qt5ClipboardSupport], [QT_LIBS="-lQt5ClipboardSupport $QT_LIBS"])
+         PKG_CHECK_MODULES([QTGRAPHICS], [Qt5GraphicsSupport], [QT_LIBS="-lQt5GraphicsSupport $QT_LIBS"])
+         PKG_CHECK_MODULES([QTCGL], [Qt5CglSupport], [QT_LIBS="-lQt5CglSupport $QT_LIBS"])
        fi
      ])
      else

jonasschnelli commented:
Seems not yet to build on gitian:

`../../corelib/tools/qbytearray.cpp:57:19: fatal error: zconf.h: No such file or directory
 #include <zconf.h>`

https://bitcoin.jonasschnelli.ch/build/616 (see build logs)

ken2812221 commented:
@jonasschnelli Seems like it's a 64-bit only issue.

ken2812221 commented:
This should fix the gitian build issue
```diff
diff --git a/depends/packages/qt.mk b/depends/packages/qt.mkindex 884a02c2b..e15a13bb0 100644
--- a/depends/packages/qt.mk
+++ b/depends/packages/qt.mk
@@ -87,6 +87,7 @@ $(package)_config_opts_linux += -fontconfig
 $(package)_config_opts_linux += -no-opengl
 $(package)_config_opts_arm_linux  = -platform linux-g++ -xplatform $(host)
 $(package)_config_opts_i686_linux  = -xplatform linux-g++-32
+$(package)_config_opts_x86_64_linux  = -xplatform linux-g++-64
 $(package)_config_opts_mingw32  = -no-opengl -xplatform win32-g++ -device-option CROSS_COMPILE="$(host)-"
 $(package)_build_env  = QT_RCC_TEST=1
 endef

fanquake commented on depends/patches/qt/qfixed-coretext.patch:
```diff
@@ -1,34 +0,0 @@
-From dbdd5f0ffbce52c8b789ed09f1aa3f1da6c02e23 Mon Sep 17 00:00:00 2001
```
Are you sure this patch is no longer required?
Compiling 99d2d3c I'm seeing:
```
compiling qtestcase.cpp
mac/qfontengine_coretext.mm:833:20: error: qualified reference to 'QFixed' is a constructor name rather than a type in this context
    return QFixed::QFixed(int(CTFontGetUnitsPerEm(ctfont)));
                   ^
1 error generated.
```

fanquake commented on depends/packages/qt.mk:
```diff
@@ -137,19 +123,17 @@ define $(package)_preprocess_cmds
   cp -f qtbase/mkspecs/macx-clang/Info.plist.app qtbase/mkspecs/macx-clang-linux/ &&\
   cp -f qtbase/mkspecs/macx-clang/qplatformdefs.h qtbase/mkspecs/macx-clang-linux/ &&\
   cp -f $($(package)_patch_dir)/mac-qmake.conf qtbase/mkspecs/macx-clang-linux/qmake.conf && \
-  patch -p1 < $($(package)_patch_dir)/mingw-uuidof.patch && \
```
Can you move all the patching back up here?

TheCharlatan replied:
Yes, moved.

sedited committed:
Ugrade Qt depends to Qt5.9.4

Depends can now be built with Qt5.9.4 , which is Qt's new long term
support version.

ken2812221 committed:
Fix depends Qt5.9.4 mac build

Apply patch from QTBUG-67286

TheCharlatan commented on depends/packages/qt.mk:
```diff
@@ -161,19 +147,22 @@ define $(package)_config_cmds
-  $(MAKE) -C src INSTALL_ROOT=$($(package)_staging_dir) $(addsuffix -install_subtargets,$(addprefix sub-,$($(package)_qt_libs))) && cd .. &&\
+  $(MAKE) -C src INSTALL_ROOT=$($(package)_staging_dir) $(addsuffix -install_subtargets,$(addprefix sub-,$($(package)_qt_libs))) && cd .. && \
   $(MAKE) -C qttools/src/linguist/lrelease INSTALL_ROOT=$($(package)_staging_dir) install_target && \
+  $(MAKE) -C qttools/src/linguist/lupdate INSTALL_ROOT=$($(package)_staging_dir) install_target && \
```
nit-picking here: is it preferable to use `&& \` or `&&\` ?

fanquake replied:
Best to just match the code around the lines you're touching. However please don't change any lines your not actually modifying just to change `&& \` to `&&\` or vice-versa.

theuni commented on depends/packages/qt.mk:
```diff
@@ -1,22 +1,20 @@
-$(package)_qt_libs=corelib network widgets gui plugins testlib
-$(package)_patches=mac-qmake.conf mingw-uuidof.patch pidlist_absolute.patch fix-xcb-include-order.patch
-$(package)_patches+=fix_qt_pkgconfig.patch fix-cocoahelpers-macos.patch qfixed-coretext.patch
+$(package)_qt_libs=corelib network widgets gui plugins testlib concurrent xml
```
Please add '-no-feature-concurrent' and remove concurrent here.

theuni replied:
Concurrent should not be needed. IIRC xml is needed, but only due to qt buildsystem bugs that I believe we patch around already.
```
$(package)_config_opts += -no-feature-concurrent
$(package)_config_opts += -no-feature-xml
```

theuni commented on depends/packages/qt.mk:
```diff
@@ -95,11 +83,11 @@ endif
 $(package)_config_opts_linux  = -qt-xkbcommon
 $(package)_config_opts_linux += -qt-xcb
 $(package)_config_opts_linux += -system-freetype
-$(package)_config_opts_linux += -no-sm
```
Untested: looks like this is "-no-feature-sessionmanager" now.

theuni commented on depends/packages/qt.mk:
```diff
@@ -61,8 +52,6 @@ $(package)_config_opts += -no-sql-psql
 $(package)_config_opts += -no-sql-sqlite2
 $(package)_config_opts += -no-use-gold-linker
-$(package)_config_opts += -no-xinput2
-$(package)_config_opts += -no-xrender
```
I believe we may have to specify some new gfx options, notice that "vnc" is now enabled for Linux.

TheCharlatan replied:
Usually this would be achieved by `no-gfx-vnc`, but Qt5.8 changed large parts of the configure. `-no-xinput2` still works, I'll add it back.

theuni commented on build-aux/m4/bitcoin_qt.m4:
```diff
@@ -360,14 +381,25 @@ AC_DEFUN([_BITCOIN_QT_FIND_STATIC_PLUGINS],[
+       if test x$bitcoin_cv_qt58 = xno; then
+         PKG_CHECK_MODULES([QTPLATFORM], [Qt5PlatformSupport], [QT_LIBS="$QTPLATFORM_LIBS $QT_LIBS"])
+       else
+         PKG_CHECK_MODULES([QTFONTDATABASE], [Qt5FontDatabaseSupport], [QT_LIBS="-lQt5FontDatabaseSupport $QT_LIBS"])
```
Are these all needed for plugins? And are they all mandatory? We don't want to fail if someone has their qt configured with, for example, -no-feature-accessibility

theuni commented on build-aux/m4/bitcoin_qt.m4:
```diff
@@ -501,10 +544,15 @@ AC_DEFUN([_BITCOIN_QT_FIND_LIBS_WITHOUT_PKGCONFIG],[
+    BITCOIN_QT_CHECK(AC_SEARCH_LIBS([png_error] ,[qtpng png],,AC_MSG_WARN([libpng not found. Assuming qt has it built-in])))
+    BITCOIN_QT_CHECK(AC_SEARCH_LIBS([pcre16_exec], [qtpcre pcre16],,AC_MSG_WARN([libpcre16 not found. Assuming qt has it built-in])))
+  else
+    BITCOIN_QT_CHECK(AC_SEARCH_LIBS([png_error] ,[qtlibpng png],,AC_MSG_WARN([libpng not found. Assuming qt has it built-in])))
```
are qtlibpng/libqtpcre2 correct? Those imply linking with: -lqtlibpng and -llibqtpcre2

TheCharlatan replied:
The libraries built are libqtlibpng.a and libqtpcre2.a

theuni commented on depends/packages/qt.mk:
```diff
@@ -137,19 +124,18 @@ define $(package)_preprocess_cmds
   echo "!host_build: QMAKE_CXXFLAGS   += $($(package)_cxxflags) $($(package)_cppflags)" >> qtbase/mkspecs/common/gcc-base.conf && \
   echo "!host_build: QMAKE_LFLAGS     += $($(package)_ldflags)" >> qtbase/mkspecs/common/gcc-base.conf && \
+  echo "QMAKE_LINK_OBJECT_MAX = 10" >> qtbase/mkspecs/win32-g++/qmake.conf &&\
+  echo "QMAKE_LINK_OBJECT_SCRIPT = object_script" >> qtbase/mkspecs/win32-g++/qmake.conf &&\
```
Why are these needed?

theuni reviewed (commented):
Left a few initial comments, great work!

fanquake commented:
After discussing with @jonasschnelli irc, need to note that this will likely bump the minimum required macOS version to 10.10.

fanquake commented:
Given that 5.9.5 has been [released](https://blog.qt.io/blog/2018/04/12/qt-5-9-5-released/), this PR can also be updated to that, over 5.9.4.

sedited committed:
Upgrade Qt depends to 5.9.6

TheCharlatan commented:
Upgraded to 5.9.5

Sjors commented:
Would this also allow cross compiling to ARM? My initial attempt in #13495 failed.

TheCharlatan commented:
@Sjors I tested it with arm-linux-gnueabihf and it has the same problem with qminimal.

TheCharlatan commented:
Rebased.

laanwj commented:
utACK 8239c0c69d7beb0b411cd684d947fec07b39483c

theuni commented:
Thanks for the rebase, taking another look.

ken2812221 commented:
utACK 5e9d4fe6448371ec3e38bad4f404d385f12355c4

fanquake commented:
Thanks @TheCharlatan:

Can this be updated to Qt [5.9.6](https://blog.qt.io/blog/2018/06/11/qt-5-9-6-released/), which contains further [bugfixes](https://bugreports.qt.io/browse/QTBUG-63772?filter=19377) from 5.9.5.

Once that's done the `qfixed-coretext.patch` can be dropped as it should have been fixed in [QTBUG-67545](https://bugreports.qt.io/browse/QTBUG-67545).

TheCharlatan commented:
Force-pushed to bump the version to Qt 5.9.6, as suggested by @fanquake .

Sjors commented on build-aux/m4/bitcoin_qt.m4:
```diff
@@ -276,35 +276,56 @@ AC_DEFUN([_BITCOIN_QT_CHECK_QT5],[
+
+
+dnl Internal. Check if the linked version of Qt was built as static libs.
+dnl Requires: Qt5. This check cannot determine if Qt4 is static.
```
Nit: no need to reference Qt4 anymore

Sjors commented on depends/packages/qt.mk:
```diff
@@ -1,22 +1,20 @@
+$(package)_suffix=opensource-src-$($(package)_version).tar.xz
 $(package)_file_name=qtbase-$($(package)_suffix)
-$(package)_sha256_hash=95f83e532d23b3ddbde7973f380ecae1bac13230340557276f75f2e37984e410
+$(package)_sha256_hash=eed620cb268b199bd83b3fc6a471c51d51e1dc2dbb5374fc97a0cc75facbe36f
```
Can confirm eed620cb268b199bd83b3fc6a471c51d51e1dc2dbb5374fc97a0cc75facbe36f matches https://download.qt.io/archive/qt/5.9/5.9.6/submodules/qtbase-opensource-src-5.9.6.tar.xz.mirrorlist

Sjors commented on depends/packages/qt.mk:
```diff
@@ -1,22 +1,20 @@
-$(package)_download_path=https://download.qt.io/archive/qt/5.7/$($(package)_version)/submodules
-$(package)_suffix=opensource-src-$($(package)_version).tar.gz
+$(package)_version=5.9.6
+$(package)_download_path=http://download.qt.io/official_releases/qt/5.9/$($(package)_version)/submodules
```
Why the http downgrade?

TheCharlatan replied:
Thanks, missed this during rebase.

Sjors commented on depends/packages/qt.mk:
```diff
@@ -29,25 +27,18 @@ $(package)_config_opts += -c++std c++11
 $(package)_config_opts += -confirm-license
 $(package)_config_opts += -dbus-runtime
 $(package)_config_opts += -hostprefix $(build_prefix)
-$(package)_config_opts += -no-alsa
```
Is the reason we can drop these lines that these options been removed from QT?

TheCharlatan replied:
The config_opts lines I removed are no longer supported by qt's configure.

Sjors commented on depends/packages/qt.mk:
```diff
@@ -95,11 +86,12 @@ endif
 $(package)_config_opts_linux += -no-opengl
 $(package)_config_opts_arm_linux  = -platform linux-g++ -xplatform $(host)
 $(package)_config_opts_i686_linux  = -xplatform linux-g++-32
+$(package)_config_opts_x86_64_linux = -xplatform linux-g++-64
```
I can try building these depends on an arm64 device soon.

Sjors replied:
`cd depends && make` doesn't build QT when run from a aarch64 host, so nvm :-)

Sjors reviewed (approved):
Tested 77fd4f8 on macOS 10.13.5 (did a `make clean-all` first). Also ran the tests.

@fanquake wrote:
> this will likely bump the minimum required macOS version to 10.10. Which probably requires further discussion.

macOS 10.10 Yosemite indeed: https://doc-snapshots.qt.io/qt5-5.9/osx.html

See #13362 for discussion.

laanwj commented:
utACK 800dea88a6ad2d8b615c1cc486da5942f6e391c2
